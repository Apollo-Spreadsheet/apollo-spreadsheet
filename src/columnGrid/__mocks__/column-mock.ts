import { Header } from '../types/header.type'

export const createColumnMock = (column?: Partial<Header>) => {
	return {
		id: 'column - ' + Math.random(),
		accessor: column?.accessor ?? 'access-' + Math.random(),
		...column,
	} as Header
}
