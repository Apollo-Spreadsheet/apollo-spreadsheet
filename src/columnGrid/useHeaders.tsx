import React, { useCallback, useMemo, useRef, useState } from 'react'
import { Header, GridHeader, NestedHeader } from './types/header.type'
import { FixedColumnWidthDictionary } from './types/fixed-column-width-dictionary'
import { StretchMode } from '../types/stretch-mode.enum'
import { insertDummyCells } from '../gridWrapper/utils/insertDummyCells'

export interface FixedColumnWidthRecord {
	totalSize: number
	mapping: FixedColumnWidthDictionary
}

export interface GetColumnAt {
	(index: number, row?: number): Header | undefined
}

interface HeadersState {
	headersData: GridHeader[][]
	dynamicColumnCount: number
	getColumnAt: GetColumnAt
}

interface Props {
	headers: Header[]
	nestedHeaders?: Array<NestedHeader[]>
	minColumnWidth: number
	stretchMode?: StretchMode
}

/**
 * useHeaders manage the column configurations, parses and validates but also expose util methods necessary for the
 * creation of the headers
 * @todo Unit tests
 * @param headers
 * @param nestedHeaders
 * @param minColumnWidth
 * @param stretchMode
 */
export function useHeaders({
	headers,
	nestedHeaders,
	minColumnWidth,
	stretchMode,
}: Props): HeadersState {
	const headersData: GridHeader[][] = useMemo(() => {
		//Detect duplicated
		const duplicateColumns = headers.filter((column, i) => {
			return headers.findIndex(d => d.id === column.id) !== i
		})

		if (duplicateColumns.length) {
			throw new Error(
				`Duplicate columns were found with ids: "${duplicateColumns
					.map(d => d.id)
					.join(', ')}" in the columns array above`,
			)
		}

		//Check and maybe validate if needed
		const transformedHeaders = headers.map(
			e =>
				({
					...e,
					colSpan: 0,
					isNested: false,
				} as GridHeader),
		)

		if (nestedHeaders) {
			//Check if any row passes the limit of spanning
			nestedHeaders.forEach((row, i) => {
				const spanSize = row.reduce((acc, e) => acc + (e.colSpan ?? 0), 0)
				if (spanSize > headers.length) {
					throw new Error(
						'Span size is bigger than the main headers size, please review the configuration at row: ' +
							i,
					)
				}
			})

			const nestedTransform = nestedHeaders.map(rows => {
				return rows.map(
					(e, i) =>
						({
							id: `nested-${i}`, //Doesn't matter that much
							accessor: '',
							title: e.title,
							colSpan: e.colSpan,
							className: e.className,
							isNested: true,
						} as GridHeader),
				)
			})

			//Here we combine with headers and insert the dummy cells
			return insertDummyCells([...nestedTransform, transformedHeaders])
		}

		//Only one level is necessary
		return insertDummyCells([transformedHeaders])
	}, [headers, nestedHeaders])

	//Stores the amount of columns that we want to calculate using the remaining width of the grid
	const dynamicColumnCount = useMemo(() => {
		return headers.filter(e => !e.width).length
	}, [headers])

	const getColumnAt = useCallback(
		(index: number, row = 0) => {
			return headersData[row]?.[index]
		},
		[headersData],
	)

	return {
		headersData,
		dynamicColumnCount,
		getColumnAt,
	}
}
