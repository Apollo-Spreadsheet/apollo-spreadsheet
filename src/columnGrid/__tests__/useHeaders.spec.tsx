import React from 'react'
import { renderHook } from '@testing-library/react-hooks'
import { useHeaders } from '../useHeaders'
import { StretchMode } from '../../types/stretch-mode.enum'
import { createColumnMock } from '../__mocks__/column-mock'
import { GridHeader } from '../types/header.type'

describe('useHeaders hook', () => {
	const consoleSpy = jest
	.spyOn(console, 'error')
	.mockImplementation(() => {
	})

	afterEach(() => {
		consoleSpy.mockClear()
	})


	it('should mount without any data', () => {
		const { result } = renderHook(() => useHeaders({
			columns: [],
			nestedHeaders: [],
			minColumnWidth: 50,
			stretchMode: StretchMode.All,
		}))
		expect(result.current.gridHeaders).toEqual([[]])
		expect(result.current.dynamicColumnCount).toBe(0)
	})

	it('should mount with a dynamic and fixed header', () => {
		const headers = [createColumnMock(), createColumnMock({ width: 200 })]
		const expectedHeadersData: GridHeader[] = headers.map(e => ({
			...e,
			isNested: false,
			colSpan: 0,
		}))
		const { result } = renderHook(() => useHeaders({
			columns: headers,
			nestedHeaders: [],
			minColumnWidth: 50,
			stretchMode: StretchMode.All,
		}))
		expect(result.current.gridHeaders).toEqual([expectedHeadersData])
		expect(result.current.dynamicColumnCount).toBe(1)
	})

	it('should throw error with duplicated header ids', () => {
		const headers = [
			createColumnMock({ id: 't1', width: 200 }),
			createColumnMock({ id: 't1', width: 200 }),
		]
		try {
			renderHook(() => useHeaders({
				columns: headers,
				nestedHeaders: [],
				minColumnWidth: 50,
				stretchMode: StretchMode.All,
			}))
		} catch (ex) {
			expect(ex).toBeTruthy()
		}
	})

	it('should throw error when span is bigger than total columns', () => {
		const headers = [createColumnMock({ width: 200 })]
		try {
			renderHook(() => useHeaders({
				columns: headers,
				nestedHeaders: [[{
					title: 'span 1',
					colSpan: 4,
				}]],
				minColumnWidth: 50,
				stretchMode: StretchMode.All,
			}))
		} catch (ex) {
			expect(ex).toBeTruthy()
		}
	})

	it('should merge with nested headers', () => {
		const headers = [createColumnMock({ id: 't1', accessor: 't1', width: 200 }), createColumnMock({
			id: 't2',
			accessor: 't2',
			width: 100,
		})]
		const { result } = renderHook(() => useHeaders({
			columns: headers,
			nestedHeaders: [[{
				title: 'Test',
				colSpan: 2,
			}]],
			minColumnWidth: 50,
			stretchMode: StretchMode.All,
		}))

		expect(result.current.gridHeaders).toMatchSnapshot()
		expect(result.current.dynamicColumnCount).toBe(0)
	})
})