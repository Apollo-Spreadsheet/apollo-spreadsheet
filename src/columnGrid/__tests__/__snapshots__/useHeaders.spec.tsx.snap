// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useHeaders hook should merge with nested headers 1`] = `
Array [
  Array [
    Object {
      "accessor": "",
      "className": undefined,
      "colSpan": 2,
      "id": "nested-0",
      "isNested": true,
      "title": "Test",
    },
    Object {
      "accessor": "",
      "className": undefined,
      "colSpan": 1,
      "dummy": true,
      "dummyFor": "colSpan",
      "first": true,
      "id": "nested-0",
      "isNested": true,
      "last": true,
      "parentRow": Object {
        "accessor": "",
        "className": undefined,
        "colSpan": 2,
        "id": "nested-0",
        "isNested": true,
        "title": "Test",
      },
      "rowSpan": undefined,
      "style": Object {
        "borderLeft": 0,
      },
      "title": "Test",
      "value": "",
    },
  ],
  Array [
    Object {
      "accessor": "t1",
      "colSpan": 0,
      "id": "t1",
      "isNested": false,
      "width": 200,
    },
    Object {
      "accessor": "t2",
      "colSpan": 0,
      "id": "t2",
      "isNested": false,
      "width": 100,
    },
  ],
]
`;
